src/main/scala/io/iohk/ethereum/utils/Riemann.scala

```diff
@@ -0,0 +1,264 @@
+          }
+        }
+        case None => {
+          log.debug("create new stdout riemann client")
+          new RiemannStdoutClient()
+        }
+      }
+    }
+    c.connect()
+    log.debug("riemann client connected")
+    c
+  }
+
+  def close: Unit = riemannClient.close()
+
+  def defaultEvent: EventDSL = {
+    val event = new EventDSL(riemannClient)
+    val microsenconds =
+      TimeUnit.MILLISECONDS.toMicros(System.currentTimeMillis());
+    event.time(microsenconds).host(hostName)
+  }
+
+  def ok(service: String): EventDSL = defaultEvent.state("ok").service(service)
+  def warning(service: String): EventDSL = defaultEvent.state("warning").service(service)
+  def error(service: String): EventDSL = defaultEvent.state("error").service(service)
+  def exception(service: String, t: Throwable): EventDSL = {
+    // Format message and stacktrace
+    val desc = new StringBuilder();
+    desc.append(t.toString())
+    desc.append("\n\n");
+    t.getStackTrace.map { e =>
+      desc.append(e)
+      desc.append("\n")
+    }
+    defaultEvent
+      .service(service)
+      .state("error")
+      .tag("exception")
+      .tag(t.getClass().getSimpleName())
+      .description(desc.toString())
+  }
+  def critical(service: String): EventDSL = defaultEvent.state("critical").service(service)
+
+}
+
+object Riemann extends Riemann with Logger {}
+
+class RiemannBatchClient(config: RiemannConfiguration)
+    extends IRiemannClient
+    with Logger {
+  private def promise[A](v: A): Promise[A] = {
+    val p: Promise[A] = new Promise()
+    p.deliver(v)
+    p
+  }
+  private def simpleMsg(event: Event) = {
+    val msg = Msg.newBuilder().addEvents(event).build()
+    promise(msg)
+  }
+  private val jsonFormatter =
+    (new FormatFactory()).createFormatter(FormatFactory.Formatter.JSON)
+
+  protected val queue: BlockingQueue[Event] = new ArrayBlockingQueue(
+    config.bufferSize)
+
+  private val client = RiemannClient.tcp(config.host, config.port)
+
+  protected def sendBatch: Unit = {
+    val batch: LinkedList[Event] = new LinkedList()
+    queue.drainTo(batch, config.batchSize)
+    try {
+      log.trace("try to send batch")
+      val p = client.sendEvents(batch)
+      client.flush()
+      val result = p.deref()
+      log.trace(s"sent batch with result: $result")
+    } catch {
+      case e: IOException =>
+        log.error(e.toString)
+        batch.asScala
+          .map { event =>
+            {
+              System.err.println(s"${jsonFormatter.printToString(event)}")
+            }
+          }
+    }
+  }
+
+  implicit val context =
+    ExecutionContext.fromExecutor(Executors.newFixedThreadPool(20))
+
+  val f = Future {
+    while (true) {
+      if (queue.size() > 0) {
+        log.trace("sending batch of Riemann events")
+        sendBatch
+        log.trace("sent batch of Riemann events")
+      } else {
+        log.trace("wait for next batch of Riemann events")
+        Thread.sleep(5000)
```

Comment:
This will happen a lot. Right after sending the last batch the whole thread will be blocked for 5 seconds - it doesn't sound too good. Can we use something non-blocking here? Perhaps `ScheduledThreadPoolExecutor.schedule` or an Actor?

Reply:
changed: we now have a `Sender` that takes care of emptying the queue and then re-schedules a `new Sender`. This starts at the same time as `client.connect()` and is closed and re-created appropriately.